client/list.c

```diff
@@ -0,0 +1,180 @@
+        return (listA);
+
+    //Compare the URL Preference to sort the URL List in descending order. 
+    TDNF_ML_URL_INFO *urlA = listA->data;
+    TDNF_ML_URL_INFO *urlB = listB->data;
+
+    if((urlA == NULL) || (urlB == NULL))
+    {
+        dwError = ERROR_TDNF_INVALID_PARAMETER;
+        BAIL_ON_TDNF_ERROR(dwError);
+    }
+
+    if (urlA->preference >= urlB->preference) 
+    {
+        mergedList = listA;
+        mergedList->next = TDNFMergeList(listA->next, listB);
+    }
+    else 
+    {
+        mergedList = listB;
+        mergedList->next = TDNFMergeList(listA, listB->next);
+    }
+
+cleanup:
+    return (mergedList);
+
+error:
+    goto cleanup;
+}
+
+/* This function is used to split the nodes of the given
+ * list into two halves,and return the two lists using the
+ * reference. If the length is odd, the extra node should
+ * go in the front list.
+ */
+void 
+TDNFFrontBackSplit(
+    TDNF_ML_LIST* currHead,
+    TDNF_ML_LIST** frontRef,
+    TDNF_ML_LIST** backRef
+    )
+{
+    if((currHead == NULL) || (currHead->next == NULL))
+    {
+        return;
+    }
+
+    TDNF_ML_LIST* slowPtr = currHead; 
+    TDNF_ML_LIST* fastPtr = currHead->next;
+
+    while (fastPtr != NULL) 
+    {
+        fastPtr = fastPtr->next;
+        if (fastPtr != NULL) 
+        {
+            slowPtr = slowPtr->next;
+            fastPtr = fastPtr->next;
+        }
+    }
+
+    *frontRef = currHead;
+    *backRef = slowPtr->next;
+    slowPtr->next = NULL;
+}
+
+void 
+TDNFSortListOnPreference(
+    TDNF_ML_LIST** headUrl
+    )
+{
+    TDNF_ML_LIST* head = *headUrl;
+    TDNF_ML_LIST* listA = NULL; 
+    TDNF_ML_LIST* listB = NULL; 
+
+    if((head == NULL) || (head->next == NULL))
+    {
+        return;
+    }
+
+    TDNFFrontBackSplit(head, &listA, &listB); 
+
+    TDNFSortListOnPreference(&listA);
+    TDNFSortListOnPreference(&listB);
+
+    *headUrl = TDNFMergeList(listA, listB); 
+
+}
+  
+/* This function is used to append the list with 
+ * new node at last.
+ */
+void TDNFAppendList(
+    TDNF_ML_LIST** head_ref,
+    void *new_data
+    )
+{
+    TDNF_ML_LIST* new_node = NULL;
+    uint32_t dwError = 0;
+
+    dwError = TDNFAllocateMemory(1, sizeof(TDNF_ML_LIST), (void**)&new_node);
+    BAIL_ON_TDNF_ERROR(dwError);
+  
+    new_node->data  = new_data;
+ 
+    if(*head_ref == NULL)
+        *head_ref = new_node;
+    else
+    {
+        TDNF_ML_LIST *last_node = *head_ref;
+        while(last_node->next != NULL)
+        {
+            last_node = last_node->next;
+        }
+        last_node->next = new_node;
+    } 
+
+cleanup:
+    return;
+
+error:
+    pr_err("Memory allocation failed for new node error: %d\n", dwError);
+    goto cleanup;
+}
+
+/* This function is used to delete the full list.
+ * Also call the respective free function to free the
+ * data ptr.
+ */
+void 
+TDNFDeleteList(
+    TDNF_ML_LIST** head_ref,
+    TDNF_ML_FREE_FUNC free_func
+    )
+{
+    TDNF_ML_LIST* current = *head_ref;
```

Comment:
check args fo NULL

Reply:
Done

Reply:
init current with NULL and assign it after argument check

Reply:
done